lib/livebook/file_system/mounter.ex

```diff
@@ -0,0 +1,176 @@
+  end
+
+  @impl GenServer
+  def handle_info({:hub_connected, hub_id}, state) do
+    {:noreply, mount_file_systems(state, hub_id)}
+  end
+
+  def handle_info({:file_system_created, file_system}, state) do
+    {:noreply, mount_file_system(state, file_system)}
+  end
+
+  def handle_info({:file_system_updated, file_system}, state) do
+    {:noreply, update_file_system(state, file_system)}
+  end
+
+  def handle_info({:file_system_deleted, file_system}, state) do
+    {:noreply, unmount_file_system(state, file_system)}
+  end
+
+  def handle_info({:hub_changed, hub_id}, state) do
+    {:noreply, mount_file_systems(state, hub_id)}
+  end
+
+  def handle_info({:hub_deleted, hub_id}, state) do
+    {:noreply, unmount_file_systems(state, hub_id)}
+  end
+
+  def handle_info(:remount, state) do
+    Process.send_after(self(), :remount, state.loop_delay)
+    {:noreply, remount_file_systems(state)}
+  end
+
+  def handle_info(_message, state), do: {:noreply, state}
+
+  defp mount_file_systems(state, hub_id) do
+    case Hubs.fetch_hub(hub_id) do
+      {:ok, hub} ->
+        file_systems = Hubs.get_file_systems(hub, hub_only: true)
+        state = put_hub(state, hub.id)
+        Enum.reduce(file_systems, state, &mount_file_system(&2, &1))
+
+      :error ->
+        state
+    end
+  end
+
+  defp remount_file_systems(state) do
+    Enum.reduce(state.hubs, state, fn {hub_id, hub_data}, acc ->
+      case Hubs.fetch_hub(hub_id) do
+        {:ok, _} -> Enum.reduce(hub_data.file_systems, acc, &update_file_system(&2, &1))
+        :error -> unmount_file_systems(acc, hub_id)
+      end
+    end)
+  end
+
+  defp unmount_file_systems(state, hub_id) do
+    if metadata = state.hubs[hub_id] do
+      metadata.file_systems
+      |> Enum.reduce(state, &unmount_file_system(&2, &1))
+      |> remove_hub(hub_id)
+    else
+      state
+    end
+  end
+
+  defp mount_file_system(state, file_system) do
+    case FileSystem.mount(file_system) do
+      :ok ->
+        broadcast({:file_system_mounted, file_system})
+        put_hub_file_system(state, file_system)
+
+      {:error, _reason} ->
+        state
+    end
+  end
+
+  defp update_file_system(state, file_system) do
+    case FileSystem.mount(file_system) do
+      :ok ->
+        broadcast({:file_system_mounted, file_system})
+        put_hub_file_system(state, file_system)
+
+      {:error, _reason} ->
+        state
+    end
+  end
+
+  defp unmount_file_system(state, file_system) do
+    case FileSystem.unmount(file_system) do
+      :ok ->
+        broadcast({:file_system_unmounted, file_system})
+        remove_hub_file_system(state, file_system)
+
+      {:error, _reason} ->
+        state
+    end
+  end
+
+  defp put_hub_file_system(state, file_system) do
+    if state.hubs[file_system.hub_id] do
+      update_in(state.hubs[file_system.hub_id], fn metadata ->
+        metadata
+        |> remove_file_system(file_system)
+        |> put_file_system(file_system)
```

Comment:
put_file_system already calls `remove_file_system`, so I think we only need this?

```suggestion
        put_file_system(metadata, file_system)
```